pymongo/uri_parser.py

```diff
@@ -35,25 +40,48 @@
 DEFAULT_PORT = 27017
 
 
+def _unquoted_percent(s):
+    """Check for unescaped percent signs
+
+    :Paramaters:
+        - `s`: A string. `s` can have things like '%25', '%2525', '%E2%85%A8'
+               and 'Ⅸ' but cannot have unquoted percent like '%foo'.
+    """
+    for i in range(len(s)):
+        if s[i] == '%':
+            sub = s[i:i+3]
+            # If unquoting yields the same string this means there was an
+            # unquoted %.
+            if unquote(sub) == sub:
+                return True
+    return False
+
 def parse_userinfo(userinfo):
     """Validates the format of user information in a MongoDB URI.
-    Reserved characters like ':', '/', '+' and '@' must be escaped
-    following RFC 3986.
+    Reserved characters that are gen-delimiters (":", "/", "?", "#", "[",
+    "]", "@") as per RFC 3986 must be escaped.
 
     Returns a 2-tuple containing the unescaped username followed
     by the unescaped password.
 
     :Paramaters:
         - `userinfo`: A string of the form <username>:<password>
+
+    .. versionchanged:: 4.0
+       Reserved characters that are sub-delimiters ("!", "$", "&", "'",
+       "(", ")", "*", "+", ",", ";", "=") as per RFC 3986 need not be
+       escaped.
```

Comment:
Add a note that '%' is no longer allowed. Also this `versionchanged:: 4.0` info should be moved to `parse_uri`. Nobody is going to be using `parse_userinfo` directly.

Reply:
Done.